### ISSUE TITLE
Authentication with token

### ISSUE THREAD

**Issue by VBobCat** [jobec/django-auth-adfs#92]
In my scenario, Django hosts only REST APIs, and user authentication in Azure is made in the frontend (Angular 8, using ADAL.js).

Is it possible to authenticate the user in Django with a JWT token sent by the frontend, instead of sending their username and password?

Thank you very much.

**Comment by jobec** [comment]
Yes, see here: https://django-auth-adfs.readthedocs.io/en/latest/rest_framework.html

**Comment by VBobCat** [comment]
Hello, @jobec , in first plance, thank you very much for your reply.

However, I wasn't able to make it work, sorry.

So, my Angular frontend sends this in order to consume the API:
```
GET http://localhost:8000/api/orientacoes/orientacoes/
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:72.0) Gecko/20100101 Firefox/72.0
Accept: application/json, text/plain, */*
Accept-Language: pt-BR,pt;q=0.8,en-US;q=0.5,en;q=0.3
Accept-Encoding: gzip, deflate
Content-Type: application/json
Authorization: Bearer eyJ(...)cTA
Origin: http://localhost:4200
Connection: keep-alive
Referer: http://localhost:4200/
```
But then it receives a redirection response to `/oauth2/login?next=/api/orientacoes/orientacoes/` and further redirection to `https://login.microsoftonline.com/[my_tenant_id]/oauth2/authorize?response_type=code&client_id=[my_client_id]&resource=api%3A%2F%2F[my_api_id]&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Foauth2%2Fcallback&state=L2FwaS9vcmllbnRhY29lcy9vcmllbnRhY29lcy8%3D&scope=openid` and this one crashes with a `Cross-Origin Request Blocked: The Same Origin Policy disallows...` blah blah.

Now I suppose these redirections should be handled server-side until completion of the authentication process, am I right?

I read the docs you linked once more and noticed my Django settings are just as indicated:
```
AUTHENTICATION_BACKENDS = [ 'django_auth_adfs.backend.AdfsAccessTokenBackend', ]
REST_FRAMEWORK = {
    'DEFAULT_PERMISSION_CLASSES': ( 'rest_framework.permissions.IsAuthenticated', ),
    'DEFAULT_AUTHENTICATION_CLASSES': ( 
        'django_auth_adfs.rest_framework.AdfsAccessTokenAuthentication',
    ), 
}
```
My settings for Azure resources are also configured insider the `AUTH_ADFS` dictionary, with the same values I passed in frontend's config.

But then I saw [the sample requisition](https://django-auth-adfs.readthedocs.io/en/latest/rest_framework.html#requesting-an-access-token) and noticed it sends username and password. Now in my case I won't collect user's credentials directly. They will login directly against Microsoft's own site, using ADAL.js (with [this Angular Wrapper](https://github.com/benbaran/adal-angular4)). So I can't send their username and password. What I can send, actually, is the JWT token that is recovered after authenticantion, and which contains user's ID. 

What should I change in order to perform users' authentication server side in this scenario?

Thank you once more for your help.

**Comment by JonasKs** [comment]
> Now I suppose these redirections should be handled server-side until completion of the authentication process, am I right?

These redirects are meant for a Django frontend (or such as your Django admin dashboard). You will have to disable your API urls to be handled by the middlewares with the `LOGIN_EXEMPT_URLS` setting. The authentication and authorization for DRF is done through the settings you wrote. 

```py
AUTH_ADFS = {
    'LOGIN_EXEMPT_URLS': [
        'api/',
        'swagger/',
        'redoc/',
        ...
    ],
```

Please provide all your `DJANGO_AUTH_ADFS` settings, so we can validate them.

> But then I saw the sample requisition and noticed it sends username and password.

This is an example on how to interact with the DRF API through Python and the requests packages. Your setup with ADAL is correct. Ensure that you send the `access_token` and not the `id_token`, though.  
Jobec wrote an excellent article on how the different oauth flows works [here](https://django-auth-adfs.readthedocs.io/en/latest/oauth2_explained.html). I'd highly recommend reading through them.

In short this is what happens to your application now:  
Request comes in with a JWT -> the request passes through all the middlewares -> None of them care about a JWT token -> denied and redirected to a login.  
What you want is this:  
Request comes in with a JWT -> the request skips the ADFS middlewares -> DRF takes over, making the request go into your default permission and authentication classes -> JWT is validated -> request goes through. 

Hope this made things more clear.

**Comment by VBobCat** [comment]
These were the settings I was using:
```
AUTH_ADFS = {
    'TENANT_ID': 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
    'CLIENT_ID': 'yyyyyyyy-yyyy-yyyy-yyyy-yyyyyyyyyyyy',
    'AUDIENCE': 'api://zzzzzzzz-zzzz-zzzz-zzzz-zzzzzzzzzzzz',
    'RELYING_PARTY_ID': 'api://zzzzzzzz-zzzz-zzzz-zzzz-zzzzzzzzzzzz',
    'MIRROR_GROUPS': True,
    'CLAIM_MAPPING': {
        'first_name': 'given_name',
        'last_name': 'family_name',
        'email': 'upn'},
    'GROUP_TO_FLAG_MAPPING': {
        'is_staff': 'wwwwwwww-wwww-wwww-wwww-wwwwwwwwwwww',
        'is_superuser': 'wwwwwwww-wwww-wwww-wwww-wwwwwwwwwwww''},
    'TIMEOUT': 15,
    'LOGIN_EXEMPT_URLS': ['^$', '^admin'],
}
```
Once I changed the last line to
```
    'LOGIN_EXEMPT_URLS': ['^$', '^admin', '^api'],
```
...it started working! I'll have to think and read the docs over and over again until I'm sure I understand what happened, but this is what I ought to do, after all.

All I can do now is to thank you guys deeply!

**Comment by JonasKs** [comment]
You’re welcome and no worries, I’m glad I could help. I’ve been in your shoes 😉

**Comment by jobec** [comment]
Add to the docs

**Comment by marlonpatrick** [comment]
@VBobCat can you explain how you can make thar works with Azure? 

For example, you register only one app in Azure ("frontend app") or you need to register two apps (one for frontend and other for backend)? How did you do that?

**Comment by marlonpatrick** [comment]
I managed to configure the following scenario: angular application + backend application (django rest framework).

The explanation is in the issue [114](https://github.com/jobec/django-auth-adfs/issues/114#issuecomment-734483460).

**Comment by mihnsen** [comment]
@VBobCat I'm struggle with "RELYING_PARTY_ID" and "AUDIENCE" value or configuration. I know it has been a while, but if you can share how you set it up, that will be helpful for me.

**Comment by VBobCat** [comment]
@marlonpatrick and @mihnsen, I'm sorry for taking long to answer, life's been such a mess...

That said, I'll try to summarise my setup:
- back-end relies on Python 3.8, Django 3.1, Django Rest Framework 3.12;
- frontend is made with Angular 11, its project folder is a subfolder of the back-end project, aside app folders generated by Django; its dist files are served by the back-end using [django-spa](https://pypi.org/project/django-spa/);

Now to settings (that worked for me):

Thou shalt follow @jobec's tutorial: https://django-auth-adfs.readthedocs.io/en/latest/azure_ad_config_guide.html and https://django-auth-adfs.readthedocs.io/en/latest/rest_framework.html, but **attention** to this: 

* You'll register `https://your.domain.com/oauth2/callback` as mobile/desktop redirect URL, as per jobec's directions *but also* `https://your.domain.com/` as web redirect URL, this will be needed by angular. 
* Don't forget to add your debugging redirect urls as well (http://localhost:8000/oauth2/callback in mobile/desktop and http://localhost:8000 and http://localhost:4200 in web)
* Enable implicit grant for both access and id tokens.
* Go back to Django and configure the proper settings (I'll use the ids from jobec's example):
```
# settings.py

AUTH_ADFS = {
    "TENANT_ID": "12345678-90ab-cdef-1234-567890abcdef",
    "CLIENT_ID": "480499d2-1f78-4a58-b7bc-03ebf8780af3",
    "RELYING_PARTY_ID": "https://examplecom.onmicrosoft.com/338d463b-b82f-490b-567890abcdef",
    "AUDIENCE": "https://examplecom.onmicrosoft.com/338d463b-b82f-490b-567890abcdef",
    "LOGIN_EXEMPT_URLS": [r"^$", r"^/$", r"^api"]  # root url must also be login exempt in order SPA files can load
}

AUTHENTICATION_BACKENDS = [
    'django_auth_adfs.backend.AdfsAccessTokenBackend',
    'django_auth_adfs.backend.AdfsAuthCodeBackend'
]


# urls.py

urlpatterns = [
    path('admin/', admin.site.urls),
    path('oauth2/', include('django_auth_adfs.urls')),
    path('api/', include(router.urls)),
]
```

Now for the Angular frontend, you'll use this: https://github.com/benbaran/adal-angular4

Do exactly as its README tells, but the config object, that must be tweaked a little, that's how I did it:

export const environment = {
  production: false,
  config: {
    tenant: "12345678-90ab-cdef-1234-567890abcdef",
    clientId: "480499d2-1f78-4a58-b7bc-03ebf8780af3",
    endpoints: {
      "https://graph.microsoft.com": "00000003-0000-0000-c000-000000000000",
      "https://your.domain.com": "api://338d463b-b82f-490b-567890abcdef"
    },
    redirectUri: "https://your.domain.com",
    postLogoutRedirectUri: "https://your.domain.com/"
}
```

I guess that makes it. It's working for me, but I can't assure it's **how it should be done**. If someone spots any flaw or weakness in this set up, I'd be most grateful if one should point it out.

Happy coding.

**Comment by afga97** [comment]
@VBobCat 

Hello

I am following your configuration but at the moment of sending the acces_token I get the following error in the backend

Could you help me achieve it?


![image](https://user-images.githubusercontent.com/41203767/104638966-daae8e80-5674-11eb-8c98-568652e8d477.png)

**Comment by JonasKs** [comment]
I’ve replied to you in #114, but see #118 for a workaround. #120 will fix that issue.
